File: scripts/initialize.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { TokenFaucet } from "../target/types/token_faucet";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import * as fs from "fs";

// Initialize the faucet with devnet USDC mint
async function main() {
  // Configure the client to use devnet
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  // Load the program
  const program = anchor.workspace.TokenFaucet as Program<TokenFaucet>;
  console.log("Program ID:", program.programId.toString());

  // Get the admin from the configured keypair
  const admin = provider.wallet.payer;
  console.log("Admin pubkey:", admin.publicKey.toString());

  // Use devnet USDC mint address
  console.log("Using devnet USDC mint...");
  const usdcMint = new anchor.web3.PublicKey(
    "7ggkvgP7jijLpQBV5GXcqugTMrc2JqDi9tiCH36SVg7A"
  );
  console.log("USDC mint address:", usdcMint.toString());
  fs.writeFileSync("usdc-mint.txt", usdcMint.toString());

  // Derive PDA addresses
  const [faucetStateAddress] = anchor.web3.PublicKey.findProgramAddressSync(
    [Buffer.from("faucet_state")],
    program.programId
  );
  console.log("Faucet state address:", faucetStateAddress.toString());

  const [faucetSolVaultAddress] = anchor.web3.PublicKey.findProgramAddressSync(
    [Buffer.from("sol_vault")],
    program.programId
  );
  console.log("Faucet SOL vault address:", faucetSolVaultAddress.toString());

  const [usdcVaultAddress] = anchor.web3.PublicKey.findProgramAddressSync(
    [Buffer.from("usdc_vault")],
    program.programId
  );
  console.log("USDC vault address:", usdcVaultAddress.toString());

  // Initialize the faucet
  console.log("Initializing faucet...");
  try {
    await program.methods
      .initialize()
      .accountsStrict({
        faucetState: faucetStateAddress,
        faucetSolVault: faucetSolVaultAddress,
        usdcMint: usdcMint,
        usdcVault: usdcVaultAddress,
        admin: admin.publicKey,
        systemProgram: anchor.web3.SystemProgram.programId,
        tokenProgram: TOKEN_PROGRAM_ID,
        rent: anchor.web3.SYSVAR_RENT_PUBKEY,
      })
      .signers([admin])
      .rpc();
    console.log("Faucet initialized successfully!");

    // Fetch and display faucet state
    const faucetState = await program.account.faucetState.fetch(
      faucetStateAddress
    );
    console.log("Faucet state:", {
      admin: faucetState.admin.toString(),
      usdcMint: faucetState.usdcMint.toString(),
      usdcVault: faucetState.usdcVault.toString(),
      bump: faucetState.bump,
      solVaultBump: faucetState.solVaultBump,
    });
  } catch (e) {
    console.error("Error initializing faucet:", e);
  }
}

main().then(
  () => process.exit(0),
  (err) => {
    console.error(err);
    process.exit(1);
  }
);

File: programs/token-faucet/src/instructions/initialize.rs
use crate::state::FaucetState;
use anchor_lang::prelude::*;
use anchor_spl::token::{Mint, Token, TokenAccount};

#[derive(Accounts)]
pub struct Initialize<'info> {
    #[account(
        init,
        payer = admin,
        space = 8 + 32 + 32 + 32 + 1 + 1, // discriminator + admin pubkey + usdc_mint + usdc_vault + bump + sol_vault_bump
        seeds = [b"faucet_state"],
        bump
    )]
    pub faucet_state: Account<'info, FaucetState>,

    #[account(
        seeds = [b"sol_vault"],
        bump,
    )]
    pub faucet_sol_vault: SystemAccount<'info>,

    pub usdc_mint: Account<'info, Mint>,

    #[account(
        init,
        payer = admin,
        seeds = [b"usdc_vault"],
        bump,
        token::mint = usdc_mint,
        token::authority = faucet_state,
    )]
    pub usdc_vault: Account<'info, TokenAccount>,

    #[account(mut)]
    pub admin: Signer<'info>,

    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
    pub rent: Sysvar<'info, Rent>,
}

pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
    let faucet_state = &mut ctx.accounts.faucet_state;
    faucet_state.admin = ctx.accounts.admin.key();
    faucet_state.usdc_mint = ctx.accounts.usdc_mint.key();
    faucet_state.usdc_vault = ctx.accounts.usdc_vault.key();
    faucet_state.bump = ctx.bumps.faucet_state;
    faucet_state.sol_vault_bump = ctx.bumps.faucet_sol_vault;

    Ok(())
}
